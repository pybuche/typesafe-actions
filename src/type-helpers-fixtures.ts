import { createAction } from './create-action';
import { createAsyncAction } from './create-async-action';
import { createCustomAction } from './create-custom-action';

/**
 * Fixtures
 */

export namespace types {
  // String Literal Types
  export const VERY_DEEP_WITH_TYPE_ONLY = 'VERY_DEEP_WITH_TYPE_ONLY';
  export const WITH_TYPE_ONLY = 'WITH_TYPE_ONLY';
  export const WITH_PAYLOAD = 'WITH_PAYLOAD';
  export const WITH_OPTIONAL_PAYLOAD = 'WITH_OPTIONAL_PAYLOAD';
  export const WITH_PAYLOAD_META = 'WITH_PAYLOAD_META';
  export const WITH_MAPPED_PAYLOAD = 'WITH_MAPPED_PAYLOAD';
  export const WITH_MAPPED_PAYLOAD_META = 'WITH_MAPPED_PAYLOAD_META';
}

export const actions = {
  deep: {
    nested: {
      withTypeOnly: createCustomAction(types.VERY_DEEP_WITH_TYPE_ONLY),
    },
  },
  withTypeOnly: createAction(types.WITH_TYPE_ONLY)<undefined>(),
  withPayload: createAction(types.WITH_PAYLOAD)<number>(),
  withOptionalPayload: createAction(types.WITH_OPTIONAL_PAYLOAD)<
    number | undefined
  >(),
  withPayloadMeta: createAction(types.WITH_PAYLOAD_META)<number, string>(),
  withMappedPayload: createAction(
    types.WITH_MAPPED_PAYLOAD,
    (payload: number) => payload
  )(),
  withMappedPayloadMeta: createAction(
    types.WITH_MAPPED_PAYLOAD_META,
    (payload: number, meta: string) => payload,
    (payload: number, meta: string) => meta
  )(),
  asyncAction: createAsyncAction(
    'FETCH_USER_REQUEST',
    'FETCH_USER_SUCCESS',
    'FETCH_USER_FAILURE'
  )<undefined, { firstName: string; lastName: string }, Error>(),
};
