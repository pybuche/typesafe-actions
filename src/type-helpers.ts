/**
 * PUBLIC API
 */

/**
 * @desc Interface for internal types augmentation
 * @example ```
 * declare module 'typesafe-actions' {
 *   export type RootAction = ActionType<typeof import('./root-action').default>;
 *   export interface Types {
 *     RootAction: RootAction;
 *   }
 * } ```
 */
export interface Types {}

/**
 * @desc Type representing Type Constant
 */
export type TypeConstant = string;

/**
 * @desc Type representing Generic Action
 */
export type Action<TType extends TypeConstant = TypeConstant> = {
  type: TType;
};

/**
 * @desc Type representing Generic ActionCreator
 */
export type ActionCreator<TType extends TypeConstant = TypeConstant> = (
  ...args: any[]
) => Action<TType>;

/**
 * @desc Type representing Generic Reducer
 */
export type Reducer<TState, TAction extends Action> = (
  state: TState | undefined,
  action: TAction
) => TState;

/**
 * @desc Action without Payload
 */
export type EmptyAction<TType extends TypeConstant> = {
  type: TType;
};

/**
 * @desc Action with only Payload
 */
export type PayloadAction<TType extends TypeConstant, TPayload> = {
  type: TType;
  payload: TPayload;
};

// /**
//  * @desc Action with only Meta
//  */
// export type MetaAction<TType extends TypeConstant, TMeta> = {
//   type: TType;
//   meta: TMeta;
// };

/**
 * @desc Action with both Payload and Meta
 */
export type PayloadMetaAction<TType extends TypeConstant, TPayload, TMeta> = {
  type: TType;
  payload: TPayload;
  meta: TMeta;
};

/**
 * @desc Action Creator producing EmptyAction
 */
export type EmptyActionCreator<TType extends TypeConstant> = () => EmptyAction<
  TType
>;

/**
 * @desc Action Creator producing PayloadAction
 */
export type PayloadActionCreator<TType extends TypeConstant, TPayload> = (
  payload: TPayload
) => PayloadAction<TType, TPayload>;

// /**
//  * @desc Action Creator producing MetaAction
//  */
// export type MetaActionCreator<TType extends TypeConstant, TMeta> = (
//   payload: undefined,
//   meta: TMeta
// ) => MetaAction<TType, TMeta>;

/**
 * @desc Action Creator producing PayloadMetaAction
 */
export type PayloadMetaActionCreator<
  TType extends TypeConstant,
  TPayload,
  TMeta
> = (
  payload: TPayload,
  meta: TMeta
) => PayloadMetaAction<TType, TPayload, TMeta>;

/**
 * @desc Type representing type getter on Action Creator instance
 */
export interface ActionCreatorTypeMetadata<TType extends TypeConstant> {
  getType?: () => TType;
}

/**
 * @desc Infers Action union-type from action-creator map object
 */
export type ActionType<
  TActionCreatorOrMap extends any
> = TActionCreatorOrMap extends ActionCreator<TypeConstant>
  ? ReturnType<TActionCreatorOrMap>
  : TActionCreatorOrMap extends Record<any, any>
  ? {
      [K in keyof TActionCreatorOrMap]: ActionType<TActionCreatorOrMap[K]>;
    }[keyof TActionCreatorOrMap]
  : TActionCreatorOrMap extends infer R // TODO: should be just never but compiler yell with circularly references itself error
  ? never
  : never;

/**
 * @desc Infers State object from reducer map object
 */
export type StateType<
  TReducerOrMap extends any
> = TReducerOrMap extends Reducer<any, any>
  ? ReturnType<TReducerOrMap>
  : TReducerOrMap extends Record<any, any>
  ? { [K in keyof TReducerOrMap]: StateType<TReducerOrMap[K]> }
  : never;

/**
 * INTERNAL API
 */

/** @private */
export type ActionCreatorBuilder<
  TType extends TypeConstant,
  TPayload extends any = undefined,
  TMeta extends any = undefined
> = [TMeta] extends [undefined]
  ? [TPayload] extends [undefined]
    ? unknown extends TPayload
      ? PayloadActionCreator<TType, TPayload>
      : unknown extends TMeta
      ? PayloadMetaActionCreator<TType, TPayload, TMeta>
      : EmptyActionCreator<TType>
    : PayloadActionCreator<TType, TPayload>
  : PayloadMetaActionCreator<TType, TPayload, TMeta>;

/** @private */
export type ActionBuilder<
  TType extends TypeConstant,
  TPayload extends any = undefined,
  TMeta extends any = undefined
> = [TMeta] extends [undefined]
  ? [TPayload] extends [undefined]
    ? unknown extends TPayload
      ? PayloadAction<TType, TPayload>
      : unknown extends TMeta
      ? PayloadMetaAction<TType, TPayload, TMeta>
      : EmptyAction<TType>
    : PayloadAction<TType, TPayload>
  : PayloadMetaAction<TType, TPayload, TMeta>;

/** @private */
export type ResolveType<T extends any> = T extends Function
  ? T
  : { [K in keyof T]: T[K] };
