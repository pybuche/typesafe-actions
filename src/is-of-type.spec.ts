import * as TH from './type-helpers';
import { isOfType } from './is-of-type';

import { actions, types } from './type-helpers-fixtures';
const {
  withTypeOnly,
  withPayload,
  withPayloadMeta,
  withMappedPayload,
  withMappedPayloadMeta,
} = actions;

/** HELPERS */

const typeOnlyAction = withTypeOnly();
const typeOnlyExpected = { type: 'WITH_TYPE_ONLY' };
const payloadAction = withPayload(2);
const payloadExpected = { type: 'WITH_PAYLOAD', payload: 2 };
const payloadMetaAction = withPayloadMeta(2, 'metaValue');
const payloadMetaExpected = {
  type: 'WITH_PAYLOAD_META',
  payload: 2,
  meta: 'metaValue',
};
const mappedPayloadAction = withMappedPayload(2);
const mappedPayloadExpected = { type: 'WITH_MAPPED_PAYLOAD', payload: 2 };
const mappedPayloadMetaAction = withMappedPayloadMeta(2, 'metaValue');
const mappedPayloadMetaExpected = {
  type: 'WITH_MAPPED_PAYLOAD_META',
  payload: 2,
  meta: 'metaValue',
};

const $action = [
  typeOnlyAction,
  payloadAction,
  payloadMetaAction,
  mappedPayloadAction,
  mappedPayloadMetaAction,
];

/** TESTS */

describe('isOfType', () => {
  describe('should work with single action-type arg', () => {
    // @dts-jest:pass:snap
    isOfType(types.WITH_TYPE_ONLY)(typeOnlyAction); // => true
    // @dts-jest:pass:snap
    isOfType(types.WITH_TYPE_ONLY, typeOnlyAction); // => true
    // @dts-jest:pass:snap
    isOfType(types.WITH_TYPE_ONLY)(payloadAction); // => false
    // @dts-jest:pass:snap
    isOfType(types.WITH_TYPE_ONLY, payloadAction); // => false
    // @dts-jest:pass:snap
    isOfType([types.WITH_TYPE_ONLY])(typeOnlyAction); // => true
    // @dts-jest:pass:snap
    isOfType([types.WITH_TYPE_ONLY], typeOnlyAction); // => true
    // @dts-jest:pass:snap
    isOfType([types.WITH_TYPE_ONLY])(payloadAction); // => false
    // @dts-jest:pass:snap
    isOfType([types.WITH_TYPE_ONLY], payloadAction); // => false
  });

  describe('should work with multiple action-type args', () => {
    // @dts-jest:pass:snap
    isOfType([types.WITH_TYPE_ONLY, types.WITH_PAYLOAD])(typeOnlyAction); // => true
    // @dts-jest:pass:snap
    isOfType([types.WITH_TYPE_ONLY, types.WITH_PAYLOAD], typeOnlyAction); // => true
    // @dts-jest:pass:snap
    isOfType([types.WITH_TYPE_ONLY, types.WITH_PAYLOAD])(payloadAction); // => true
    // @dts-jest:pass:snap
    isOfType([types.WITH_TYPE_ONLY, types.WITH_PAYLOAD], payloadAction); // => true
    // @dts-jest:pass:snap
    isOfType([types.WITH_TYPE_ONLY, types.WITH_PAYLOAD])(mappedPayloadAction); // => false
    // @dts-jest:pass:snap
    isOfType([types.WITH_TYPE_ONLY, types.WITH_PAYLOAD], mappedPayloadAction); // => false
  });

  describe('should correctly assert for an array with 1 arg', () => {
    const actual = $action.filter(isOfType(types.WITH_TYPE_ONLY));
    // @dts-jest:pass:snap
    actual; // => [typeOnlyExpected]
  });

  describe('should correctly assert for an array with 2 args', () => {
    const actual = $action.filter(
      isOfType([types.WITH_TYPE_ONLY, types.WITH_PAYLOAD])
    );
    // @dts-jest:pass:snap
    actual; // => [typeOnlyExpected, payloadExpected]
  });

  describe('should correctly assert for an array with 3 args', () => {
    const actual = $action.filter(
      isOfType([
        types.WITH_TYPE_ONLY,
        types.WITH_PAYLOAD,
        types.WITH_PAYLOAD_META,
      ])
    );
    // @dts-jest:pass:snap
    actual; // => [typeOnlyExpected,payloadExpected,payloadMetaExpected]
  });

  describe('should correctly assert for an array with 4 args', () => {
    const actual = $action.filter(
      isOfType([
        types.WITH_TYPE_ONLY,
        types.WITH_PAYLOAD,
        types.WITH_PAYLOAD_META,
        types.WITH_MAPPED_PAYLOAD,
      ])
    );
    // @dts-jest:pass:snap
    actual; // => [typeOnlyExpected,payloadExpected,payloadMetaExpected,mappedPayloadExpected]
  });

  describe('should correctly assert for an array with 4 args', () => {
    const actual = $action.filter(
      isOfType([
        types.WITH_TYPE_ONLY,
        types.WITH_PAYLOAD,
        types.WITH_PAYLOAD_META,
        types.WITH_MAPPED_PAYLOAD,
        types.WITH_MAPPED_PAYLOAD_META,
      ])
    );
    // @dts-jest:pass:snap
    actual; // => [typeOnlyExpected,payloadExpected,payloadMetaExpected,mappedPayloadExpected,mappedPayloadMetaExpected]
  });

  describe('should correctly assert type with "any" action', () => {
    const action: any = withMappedPayload(1234);
    if (
      isOfType(
        [types.WITH_MAPPED_PAYLOAD, types.WITH_MAPPED_PAYLOAD_META],
        action
      )
    ) {
      // @dts-jest:pass:snap
      action; // => {"payload": 1234, "type": "WITH_MAPPED_PAYLOAD"}
    }
    if (
      isOfType([types.WITH_MAPPED_PAYLOAD, types.WITH_MAPPED_PAYLOAD_META])(
        action
      )
    ) {
      // @dts-jest:pass:snap
      action; // => {"payload": 1234, "type": "WITH_MAPPED_PAYLOAD"}
    }
  });
});
