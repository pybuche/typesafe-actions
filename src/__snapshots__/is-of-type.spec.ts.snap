// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action 1`] = `"any"`;

exports[`action 2`] = `"any"`;

exports[`actual 1`] = `"TH.EmptyAction<\\"WITH_TYPE_ONLY\\">[]"`;

exports[`actual 2`] = `"(TH.EmptyAction<\\"WITH_TYPE_ONLY\\"> | TH.PayloadAction<\\"WITH_PAYLOAD\\", number>)[]"`;

exports[`actual 3`] = `"(TH.EmptyAction<\\"WITH_TYPE_ONLY\\"> | TH.PayloadAction<\\"WITH_PAYLOAD\\", number> | TH.PayloadMetaAction<\\"WITH_PAYLOAD_META\\", number, string>)[]"`;

exports[`actual 4`] = `"(TH.EmptyAction<\\"WITH_TYPE_ONLY\\"> | TH.PayloadAction<\\"WITH_PAYLOAD\\", number> | TH.PayloadAction<\\"WITH_MAPPED_PAYLOAD\\", number> | TH.PayloadMetaAction<\\"WITH_PAYLOAD_META\\", number, string>)[]"`;

exports[`actual 5`] = `"(TH.EmptyAction<\\"WITH_TYPE_ONLY\\"> | TH.PayloadAction<\\"WITH_PAYLOAD\\", number> | TH.PayloadAction<\\"WITH_MAPPED_PAYLOAD\\", number> | TH.PayloadMetaAction<\\"WITH_PAYLOAD_META\\", number, string> | TH.PayloadMetaAction<\\"WITH_MAPPED_PAYLOAD_META\\", number, string>)[]"`;

exports[`isOfType([types.WITH_TYPE_ONLY, types.WITH_PAYLOAD])(mappedPayloadAction) 1`] = `"boolean"`;

exports[`isOfType([types.WITH_TYPE_ONLY, types.WITH_PAYLOAD])(payloadAction) 1`] = `"boolean"`;

exports[`isOfType([types.WITH_TYPE_ONLY, types.WITH_PAYLOAD])(typeOnlyAction) 1`] = `"boolean"`;

exports[`isOfType([types.WITH_TYPE_ONLY, types.WITH_PAYLOAD], mappedPayloadAction) 1`] = `"boolean"`;

exports[`isOfType([types.WITH_TYPE_ONLY, types.WITH_PAYLOAD], payloadAction) 1`] = `"boolean"`;

exports[`isOfType([types.WITH_TYPE_ONLY, types.WITH_PAYLOAD], typeOnlyAction) 1`] = `"boolean"`;

exports[`isOfType([types.WITH_TYPE_ONLY])(payloadAction) 1`] = `"boolean"`;

exports[`isOfType([types.WITH_TYPE_ONLY])(typeOnlyAction) 1`] = `"boolean"`;

exports[`isOfType([types.WITH_TYPE_ONLY], payloadAction) 1`] = `"boolean"`;

exports[`isOfType([types.WITH_TYPE_ONLY], typeOnlyAction) 1`] = `"boolean"`;

exports[`isOfType(types.WITH_TYPE_ONLY)(payloadAction) 1`] = `"boolean"`;

exports[`isOfType(types.WITH_TYPE_ONLY)(typeOnlyAction) 1`] = `"boolean"`;

exports[`isOfType(types.WITH_TYPE_ONLY, payloadAction) 1`] = `"boolean"`;

exports[`isOfType(types.WITH_TYPE_ONLY, typeOnlyAction) 1`] = `"boolean"`;
